File: src/lib/sounds.ts
import catalog from "./sounds/catalog.json";

export type SoundKind = "bed" | "shot";

export interface SoundVariant {
  id: number;
  url: string;
  seconds: number;
  channels: number;
  name: string;
  user: string;
  page: string;
  license: string;
}

export interface SoundTag {
  tag: string;
  kind: SoundKind;
  label: string;
  gain: number;
  variants: SoundVariant[];
}

export interface SoundCatalog {
  version: number;
  generatedAt: string;
  source: string;
  tags: SoundTag[];
}

export const SOUND_CATALOG = catalog as SoundCatalog;

export const SOUND_TAGS = SOUND_CATALOG.tags.filter((t) => t.variants.length > 0);

const byTag = new Map(SOUND_TAGS.map((t) => [t.tag, t]));

export const soundTag = (tag: string) => byTag.get(tag);

export const hasSounds = SOUND_TAGS.length > 0;

export function soundMenu(kind: SoundKind) {
  return SOUND_TAGS.filter((t) => t.kind === kind)
    .map((t) => `${t.tag} (${t.label.toLowerCase()})`)
    .join(", ");
}

export function pickVariant(tag: string, nth = 0): SoundVariant | null {
  const entry = byTag.get(tag);
  if (!entry || !entry.variants.length) return null;
  return entry.variants[Math.abs(nth) % entry.variants.length];
}

export function resolveTag(tag: string, fallback?: string): string | null {
  const clean = (v: string) => v.trim().toLowerCase().replace(/[\s-]+/g, "_");
  for (const candidate of [tag, fallback]) {
    if (!candidate) continue;
    const key = clean(candidate);
    if (byTag.has(key)) return key;
    const loose = SOUND_TAGS.find((t) => t.tag.startsWith(key) || key.startsWith(t.tag));
    if (loose) return loose.tag;
  }
  return null;
}

export function soundCredits(tags: string[]) {
  const seen = new Set<string>();
  const out: { name: string; user: string; page: string }[] = [];
  for (const tag of tags) {
    for (const v of byTag.get(tag)?.variants ?? []) {
      if (seen.has(v.page)) continue;
      seen.add(v.page);
      out.push({ name: v.name, user: v.user, page: v.page });
    }
  }
  return out;
}

File: android/app/src/main/java/com/nigixmosha/app/engine/Sounds.java
package com.nigixmosha.app.engine;

import android.content.Context;

import com.google.gson.Gson;
import com.nigixmosha.app.engine.model.Types;

import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.Reader;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

public final class Sounds {
    private static Sounds instance;

    private final List<Types.SoundTag> tags = new ArrayList<>();
    private final Map<String, Types.SoundTag> byTag = new HashMap<>();

    private Sounds(Types.SoundCatalog catalog) {
        if (catalog != null && catalog.tags != null) {
            for (Types.SoundTag tag : catalog.tags) {
                if (tag == null || tag.tag == null || tag.variants == null || tag.variants.isEmpty()) continue;
                tags.add(tag);
                byTag.put(tag.tag, tag);
            }
        }
    }

    public static synchronized Sounds get() {
        if (instance == null) instance = new Sounds(null);
        return instance;
    }

    public static synchronized void load(Context context) {
        if (instance != null) return;
        try {
            load(context.getAssets().open("sounds.json"));
        } catch (Exception e) {
            instance = new Sounds(null);
        }
    }

    public static synchronized void load(InputStream stream) {
        Types.SoundCatalog catalog = null;
        try (Reader reader = new InputStreamReader(stream, StandardCharsets.UTF_8)) {
            catalog = new Gson().fromJson(reader, Types.SoundCatalog.class);
        } catch (Exception ignored) {
        }
        instance = new Sounds(catalog);
    }

    public boolean isEmpty() {
        return tags.isEmpty();
    }

    public Types.SoundTag tag(String id) {
        return id == null ? null : byTag.get(id);
    }

    public String menu(String kind) {
        StringBuilder sb = new StringBuilder();
        for (Types.SoundTag tag : tags) {
            if (!tag.kind.equals(kind)) continue;
            if (sb.length() > 0) sb.append(", ");
            sb.append(tag.tag).append(" (").append(tag.label.toLowerCase(Locale.ROOT)).append(")");
        }
        return sb.toString();
    }

    public Types.SoundVariant variant(String id, int nth) {
        Types.SoundTag tag = tag(id);
        if (tag == null || tag.variants.isEmpty()) return null;
        return tag.variants.get(Math.abs(nth) % tag.variants.size());
    }

    public String resolve(String wanted, String fallback) {
        for (String candidate : new String[]{wanted, fallback}) {
            if (candidate == null) continue;
            String key = candidate.trim().toLowerCase(Locale.ROOT).replaceAll("[\\s-]+", "_");
            if (key.isEmpty()) continue;
            if (byTag.containsKey(key)) return key;
            for (Types.SoundTag tag : tags) {
                if (tag.tag.startsWith(key) || key.startsWith(tag.tag)) return tag.tag;
            }
        }
        return null;
    }
}
